File: src/p0784/cpp/solution.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;


class Solution {
public:
    vector<string> letterCasePermutation(string &s) {
        vector<string> result;
        dfs(s, "", result);
        return result;
    }

private:
    void dfs(const string &s, const string &t, vector<string> &result) {
        if (s.size() <= t.size()) {
            result.push_back(t); return;
        }
        auto i = t.size();
        if (isalpha(s[i])) {
            string lower(t); lower.push_back(tolower(s[i]));
            dfs(s, lower, result);
            string upper(t); upper.push_back(toupper(s[i]));
            dfs(s, upper, result);
        } else {
            string other(t); other.push_back(s[i]);
            dfs(s, other, result);
        }
    }
};

File: src/p0784/java/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> letterCasePermutation(String s) {
        List<String> result = new ArrayList<>();
        dfs(s, "", result);
        return result;
    }
    private void dfs(String s, String t, List<String> result) {
        if (s.length() <= t.length()) {
            result.add(t); return;
        }
        int i = t.length();
        if (Character.isLetter(s.charAt(i))) {
            String lower = t.concat(s.substring(i, i + 1).toLowerCase());
            dfs(s, lower, result);
            String upper = t.concat(s.substring(i, i + 1).toUpperCase());
            dfs(s, upper, result);
        } else {
            String other = t.concat(s.substring(i, i + 1));
            dfs(s, other, result);
        }
    }
}
